servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -1534,11 +1543,37 @@ void deleteDomain(ResourceContext ctx, String auditRef, String domainName, Strin
             groupMemberConsistencyCheck(domainName, group.getName(), true, caller);
         }
 
-        // consistency checks are ok so we can go ahead and delete the domain
+        // consistency checks are ok. Now make sure no service is dependent on the domain
+
+        verifyNoServiceDependenciesOnDomain(domainName, caller);
+
+        // no service is dependent on the domain, we can go ahead and delete the domain
 
         dbService.executeDeleteDomain(ctx, domainName, auditRef, caller);
     }
 
+    private void verifyNoServiceDependenciesOnDomain(String domainName, String caller) {
+        ServiceIdentityList serviceIdentityList = dbService.listServiceDependencies(domainName);
+        if (serviceIdentityList.getNames() != null && !serviceIdentityList.getNames().isEmpty()) {
+            StringBuilder msgBuilder = new StringBuilder("Remove domain '");
+            msgBuilder.append(domainName);
+            msgBuilder.append("' dependency from the following service(s):");
+            msgBuilder.append(String.join(", ", serviceIdentityList.getNames()));
+            throw ZMSUtils.forbiddenError(msgBuilder.toString(), caller);
+        }
+    }
+
+    private void verifyNoDomainDependenciesOnService(String serviceName, String caller) {
+        DomainList domainList = dbService.listDomainDependencies(serviceName);
```

Comment:
I've actually had this check and removed it because of the potential bug listed here: https://github.com/AthenZ/athenz/pull/1758#discussion_r795955497

Reply:
what you've described is a failure on our (athenz system administrator) part. why would be delete a service from a role without first checking to see if they're any dependencies. and if we end up with dangling dependencies, then the system admin will re-add the role, clean up and then follow through by removing the service.

Reply:
Fixed